src/Features/Core/Portable/Diagnostics/DiagnosticService.cs

```diff
@@ -205,8 +206,19 @@ private void OnCleared(object sender, EventArgs e)
         }
 
         public ImmutableArray<DiagnosticData> GetDiagnostics(
-            Workspace workspace, ProjectId projectId, DocumentId documentId, object id, bool includeSuppressedDiagnostics, CancellationToken cancellationToken)
+            Workspace workspace,
+            ProjectId projectId,
+            DocumentId documentId,
+            object id,
+            bool includeSuppressedDiagnostics,
+            bool forPullDiagnostics,
+            CancellationToken cancellationToken)
         {
+            // If this is a pull client, but pull diagnostics is not on, then they get nothing.  Similarly, if this is a
+            // push client and pull diagnostics are on, they get nothing.
+            if (forPullDiagnostics != workspace.Options.GetOption(InternalDiagnosticsOptions.LspPullDiagnostics))
+                return ImmutableArray<DiagnosticData>.Empty;
```

Comment:
this is one of 3 main entrypoints that now ensures the right data flows out depending on the caller's needs.  If the caller doesn't match the current diagnostics-mode we're in (either 'push' or 'pull') they get nothing back.  This allows DiagnosticService to still be the centralized authority that everyone uses.